arbor/swcio.cpp

```diff
@@ -1,213 +1,202 @@
+{}
 
-    if (r.r < 0) {
-        return "negative radii are not allowed";
-    }
+swc_irregular_id::swc_irregular_id(int record_id):
+    swc_error("SWC record id not numbered consecutively from 1", record_id)
+{}
 
-    return nullptr;
-}
+swc_spherical_soma::swc_spherical_soma(int record_id):
+    swc_error("SWC with spherical somata are not supported", record_id)
+{}
 
+bad_swc_data::bad_swc_data(int record_id):
+    swc_error("Cannot interpret bad SWC data", record_id)
+{}
 
-bool swc_record::is_consistent() const {
-    return swc_record_error(*this)==nullptr;
-}
+// Record I/O
 
-void swc_record::assert_consistent() const {
-    const char* error = swc_record_error(*this);
-    if (error) {
-        throw swc_error(error);
-    }
+std::ostream& operator<<(std::ostream& out, const swc_record& record) {
+    io::save_ios_flags save(out);
+
+    out.precision(std::numeric_limits<double>::digits10+2);
+    return out << record.id << ' ' << record.tag << ' '
+               << record.x  << ' ' << record.y   << ' ' << record.z << ' ' << record.r
+               << record.parent_id << '\n';
 }
 
-static bool parse_record(const std::string& line, swc_record& record) {
-    std::istringstream is(line);
-    swc_record r;
+std::istream& operator>>(std::istream& in, swc_record& record) {
+    std::string line;
+    if (!getline(in, line, '\n')) return in;
 
-    is >> r.id >> r.tag >> r.x >> r.y >> r.z >> r.r >> r.parent_id;
-    if (is) {
-        // Convert to zero-based, leaving parent_id as-is if -1
-        --r.id;
-        if (r.parent_id>=0) {
-            --r.parent_id;
-        }
+    swc_record r;
+    std::istringstream s(line);
+    s >> r.id >> r.tag >> r.x >> r.y >> r.z >> r.r >> r.parent_id;
+    if (s) {
         record = r;
-        return true;
     }
-    return false;
-}
+    else {
+        in.setstate(std::ios_base::failbit);
+    }
 
-static bool is_comment(const std::string& line) {
-    auto pos = line.find_first_not_of(" \f\n\r\t\v");
-    return pos==std::string::npos || line[pos]=='#';
+    return in;
 }
 
-std::istream& operator>>(std::istream& is, swc_record& record) {
-    std::string line;
+// Parse SWC format data (comments and sequence of SWC records).
 
-    while (is) {
-        std::getline(is, line);
-        if (!is) {
-            break;
+static std::vector<swc_record> sort_and_validate_swc(std::vector<swc_record> records, swc_mode mode) {
+    if (records.empty()) return {};
+
+    std::unordered_set<int> seen;
+    std::size_t n_rec = records.size();
+    int first_id = records[0].id;
+    int first_tag = records[0].tag;
+
+    if (records.size()<2) {
+        throw swc_spherical_soma(first_id);
+    }
+
+    for (std::size_t i = 0; i<n_rec; ++i) {
+        swc_record& r = records[i];
+
+        if (r.parent_id>=r.id) {
+            throw swc_record_precedes_parent(r.id);
         }
 
-        if (is_comment(line)) {
-            continue;
+        if (mode==swc_mode::strict && r.id != (int)i+1) {
```

Comment:
If would prefer to see `if (mode&swc_mode::contiguous && ...)`, so that we can pick and choose the restrictions, as required (unfortunately) by SONATA.

Reply:
Can't avoid special handling of spherical somata: I think in this special handling one should perform any additional checks required by e.g. SONATA.